mmgen/models/architectures/sngan_proj/generator_discriminator.py

```diff
@@ -168,119 +182,115 @@ class SNGANDiscriminator(nn.Module):
+        self.from_rgb_cfg = deepcopy(self._default_from_rgb_cfg)
+        if from_rgb_cfg is not None:
+            self.from_rgb_cfg.update(from_rgb_cfg)
+        self.from_rgb_cfg['act_cfg'] = act_cfg
+        self.from_rgb_cfg['with_spectral_norm'] = with_spectral_norm
+
+        # load and update blocks_cfg
+        self.blocks_cfg = deepcopy(self._default_blocks_cfg)
+        if blocks_cfg is not None:
+            self.blocks_cfg.update(blocks_cfg)
+        self.blocks_cfg['act_cfg'] = act_cfg
+        self.blocks_cfg['with_spectral_norm'] = with_spectral_norm
+
+        # get config for channel factor list coresponding to the input_scale
+        # channel_factor_list:
         channel_cfg = self._defualt_channels_cfg if channel_cfg is None \
             else channel_cfg
         if isinstance(channel_cfg, dict):
             if input_scale not in channel_cfg:
-                raise KeyError(
-                    '`output_scale={} is not found in `channel_cfg`, '
-                    'only support configs for {}'.format(
-                        input_scale, [chn for chn in channel_cfg.keys()]))
+                raise KeyError(f'`input_scale={input_scale} is not found in '
+                               '`channel_cfg`, only support configs for '
+                               f'{[chn for chn in channel_cfg.keys()]}')
             self.channel_factor_list = channel_cfg[input_scale]
 
+        # get config for downsample coresponding to the input_scale
+        # downsample_cfg: whether apply downsample in this block
         downsample_cfg = self._defualt_downsample_list_cfg \
             if downsample_cfg is None else downsample_cfg
         if isinstance(downsample_cfg, dict):
             if input_scale not in downsample_cfg:
-                raise KeyError(
-                    '`output_scale={} is not found in `downsample_cfg`, '
-                    'only support configs for {}'.format(
-                        input_scale, [chn for chn in downsample_cfg.keys()]))
+                raise KeyError(f'`output_scale={input_scale} is not found in '
+                               '`downsample_cfg`, only support configs for '
+                               f'{[chn for chn in downsample_cfg.keys()]}')
             self.downsample_list = downsample_cfg[input_scale]
 
         if len(downsample_cfg) != len(channel_cfg):
-            raise ValueError('TODO')
+            raise ValueError(
+                '`downsample_cfg` should have same length with `channel_cfg`, '
+                f'but receive {len(downsample_cfg)} and {len(channel_cfg)}.')
 
-        self.from_rgb = SNGAN_DisHeadResBlock(input_channels, base_channels,
-                                              **self.head_cfg)
+        self.from_rgb = build_from_cfg(
+            self.from_rgb_cfg, MODULES,
+            dict(in_channels=input_channels, out_channels=base_channels))
 
-        self.res_block = nn.ModuleList()
+        self.blocks = nn.ModuleList()
```

Comment:
In `Genernator`, you name it `conv_blocks` and it is better to keep the consistency.